File: project3/Extension/language_comparison.js
/**
 * File: compare_js.js
 * Author: Simon Lartey
 * Date: 10/19/2025
 *
 * Description:
 * JavaScript version of code demonstrating syntax, scoping, and function behavior.
 * Mirrors the Python example for explicit comparison.
 */

// ---------------------------
// 1️⃣ Variable Declaration
// ---------------------------
let x = 42;
let y = "hello";
let z = [1, 2, 3];

console.log("JavaScript Variable Types:");
console.log("x =", x, "type:", typeof x);
console.log("y =", y, "type:", typeof y);
console.log("z =", z, "type:", typeof z);
console.log("\n");


// ---------------------------
// 2️⃣ Scoping Rules (block + lexical)
// ---------------------------
let globalVar = "global";

function outerFunction() {
    let outerVar = "outer";
    
    function innerFunction() {
        let innerVar = "inner";
        console.log(`Accessing ${globalVar} from global scope`);
        console.log(`Accessing ${outerVar} from enclosing scope`);
        console.log(`Accessing ${innerVar} from local scope`);
    }

    innerFunction();
}

outerFunction();
console.log("\n");


// ---------------------------
// 3️⃣ Control Flow
// ---------------------------
let n = 5;
if (n > 10) {
    console.log("n is greater than 10");
} else if (n === 5) {
    console.log("n is exactly 5");
} else {
    console.log("n is less than 10");
}
console.log();

for (let i = 0; i < 3; i++) {
    console.log(`for loop count: ${i}`);
}
console.log();

while (n > 0) {
    console.log(`while loop count: ${n}`);
    n--;
}
console.log();


// ---------------------------
// 4️⃣ Function as First-Class Object
// ---------------------------
function greet(name) {
    return `Hello, ${name}!`;
}

// Assign function to variable
const welcome = greet;
console.log(welcome("Simon"));

// Pass function as argument
function execute(func, arg) {
    console.log("Executing passed function...");
    console.log(func(arg));
}

execute(greet, "Joyce");

// Return function from function
function multiplier(factor) {
    return function(n) {
        return n * factor;
    };
}

const double = multiplier(2);
console.log("Double of 10 is:", double(10));
console.log();


// ---------------------------
// 5️⃣ Built-in Types and Operators
// ---------------------------
let a = 10, b = 3;
console.log(`Addition: ${a + b}`);
console.log(`Subtraction: ${a - b}`);
console.log(`Multiplication: ${a * b}`);
console.log(`Division: ${a / b}`);
console.log(`Modulus: ${a % b}`);
console.log(`Exponentiation: ${a ** b}`);
console.log();

File: project3/Extension/language_comparison.py
"""
File: compare_js.py
Author: Simon Lartey
Date: 10/19/2025

Description:
Python version of code demonstrating syntax, scoping, and function behavior.
This script mirrors the JavaScript examples to explicitly compare the two languages.
"""

# ---------------------------
# 1️⃣ Variable Declaration
# ---------------------------
x = 42          # Python infers type dynamically
y = "hello"
z = [1, 2, 3]

print("Python Variable Types:")
print("x =", x, "type:", type(x))
print("y =", y, "type:", type(y))
print("z =", z, "type:", type(z))
print()


# ---------------------------
# 2️⃣ Scoping Rules (LEGB)
# ---------------------------
global_var = "global"

def outer_function():
    outer_var = "outer"

    def inner_function():
        inner_var = "inner"
        print(f"Accessing {global_var} from global scope")
        print(f"Accessing {outer_var} from enclosing scope")
        print(f"Accessing {inner_var} from local scope")
    
    inner_function()

outer_function()
print()


# ---------------------------
# 3️⃣ Control Flow
# ---------------------------
n = 5
if n > 10:
    print("n is greater than 10")
elif n == 5:
    print("n is exactly 5")
else:
    print("n is less than 10")
print()

for i in range(3):
    print(f"for loop count: {i}")
print()

while n > 0:
    print(f"while loop count: {n}")
    n -= 1
print()


# ---------------------------
# 4️⃣ Function as First-Class Object
# ---------------------------
def greet(name):
    return f"Hello, {name}!"

# Assign function to variable
welcome = greet
print(welcome("Simon"))

# Pass function as argument
def execute(func, arg):
    print("Executing passed function...")
    print(func(arg))

execute(greet, "Joyce")

# Return function from function
def multiplier(factor):
    def multiply(n):
        return n * factor
    return multiply

double = multiplier(2)
print("Double of 10 is:", double(10))
print()


# ---------------------------
# 5️⃣ Built-in Types and Operators
# ---------------------------
a, b = 10, 3
print(f"Addition: {a + b}")
print(f"Subtraction: {a - b}")
print(f"Multiplication: {a * b}")
print(f"Division: {a / b}")
print(f"Modulus: {a % b}")
print(f"Exponentiation: {a ** b}")
print()
